packages/firebase_database/lib/src/database_reference.dart

```diff
@@ -125,10 +128,65 @@ class DatabaseReference extends Query {
   ///
   /// remove() is equivalent to calling set(null)
   Future<Null> remove() => set(null);
+
+  /// Determine the next transaction key to be used. 0 is used if there are
+  /// no existing keys, otherwise the last key is incremented by 1.
+  int _getNextTransactionKey() {
+    if (FirebaseDatabase._transactions.isEmpty) {
+      return 0;
+    }
+    final int lastKey = FirebaseDatabase._transactions.keys.last;
+    return lastKey + 1;
+  }
+
+  /// Performs an optimistic-concurrency transactional update to the data at
+  /// this Firebase Database location.
+  Future<Null> runTransaction(TransactionHandler transactionHandler,
+      [int transactionTimeout = _transactionTimeout]) async {
+    if (transactionTimeout < 0) {
+      throw new ArgumentError(
+          'Transaction timeout ($transactionTimeout) cannot be less than zero.');
+    }
+    final int transactionKey = _getNextTransactionKey();
+    FirebaseDatabase._transactions[transactionKey] = transactionHandler;
+
+    final Map<String, dynamic> completion = await _database._channel
```

Comment:
Make `runTransaction` be a normal (non-async) method. 

Create a `Completer<TransactionResult> completer = new Completer<TransactionResult>` here. Call `then` on the result of `invokeMethod` to get the results of the transaction and use it to call `completer.complete()` with a `TransactionResult` object in the callback. 

Finally, return `completer.future` after the call to `invokeMethod`.

Reply:
Done!